src/librustc_yk_sections/mir_cfg.rs

```diff
@@ -64,43 +150,192 @@ pub fn generate_tir<'a, 'tcx, 'gcx>(
+                    s.push(*y);
+                }
+            }
+
+            // The second stage of the algorithm needs the children nodes in the dominator tree.
+            let mut children = Vec::new();
+            for (b, _) in self.mir.basic_blocks().iter_enumerated() {
+                let b = BasicBlock::from_u32(b.as_u32());
+                if self.doms.is_dominated_by(b, n) && b != n {
+                    children.push(b);
+                    // Force the frontier of `b` into `self.df`. Doing this here avoids a
+                    // simultaneous mutable + immutable borrow of `self` in the final stage of the
+                    // algorithm.
+                    self.get(b);
+                }
+            }
+
+            // Compute what Appel calls `DF_{up}[c]` for each dominator tree child `c` of `n`.
+            let mut df_up_cs = Vec::new();
+            for c in children {
+                for w in self.df[c].as_ref().unwrap() {
+                    if n == *w || !self.doms.is_dominated_by(*w, n) {
+                        df_up_cs.push(*w);
+                    }
+                }
+            }
+            s.extend(df_up_cs);
+
+            self.df[n] = Some(s);
+        }
+
+        self.df[n].as_ref().unwrap()
+    }
+}
+
+/// This struct deals with inserting PHI nodes into the initial pre-SSA TIR pack.
+///
+/// See the bottom of Page 406 of the 2nd edition of the Appel book mention above.
+struct PhiInserter<'a, 'tcx> {
+    mir: &'a Mir<'tcx>,
+    pack: ykpack::Pack,
+    def_sites: Vec<Vec<BasicBlock>>,
+}
+
+impl<'a, 'tcx> PhiInserter<'a, 'tcx> {
+    fn new(mir: &'a Mir<'tcx>, pack: ykpack::Pack, def_sites: Vec<Vec<BasicBlock>>) -> Self {
+        Self {
+            mir,
+            pack,
+            def_sites,
+        }
+    }
+
+    /// Insert PHI nodes, returning the mutated pack.
+    fn pack(mut self) -> ykpack::Pack {
+        let mut df = DominatorFrontiers::new(self.mir);
+
+        // We first need a mapping from block to the variables it defines. Appel calls this
+        // `A_{orig}`. We can derive this from our definition sites.
+        let (a_orig, num_blks) = {
+            let ykpack::Pack::Mir(ykpack::Mir{ref blocks, ..}) = self.pack;
+            let mut a_orig: IndexVec<BasicBlock, Vec<TirVarIndex>> = IndexVec::with_capacity(blocks.len());
+            a_orig.resize(blocks.len(), Vec::default());
+            for (a, def_blks) in self.def_sites.iter().enumerate() {
+                for bb in def_blks {
+                    a_orig[*bb].push(u32::try_from(a).unwrap())
+                }
+            }
+
+            (a_orig, blocks.len())
+        };
+
+        let mut a_phi: Vec<Vec<TirVarIndex>> = Vec::with_capacity(num_blks);
+        a_phi.resize(num_blks, Vec::new());
+        for (a, def_sites) in self.def_sites.iter().enumerate() {
+            let mut w = def_sites.clone();
+            while w.len() > 0 {
+                let n = w.pop().unwrap();
```

Comment:
The intermediate variable seems of little use?

Reply:
Solely to pair with the original algorithm:

![image](https://user-images.githubusercontent.com/604955/54441837-71d2f400-4735-11e9-90bf-15db712a4ff2.png)

Reply:
I think that text needs to be bigger ;)

Anyway, fair enough.